src/app/clusters/time-synchronization-server/time-synchronization-server.cpp

```diff
@@ -0,0 +1,980 @@
+
+#include <app-common/zap-generated/attributes/Accessors.h>
+#include <app-common/zap-generated/cluster-objects.h>
+#include <app-common/zap-generated/ids/Attributes.h>
+#include <app-common/zap-generated/ids/Clusters.h>
+#include <app/AttributeAccessInterface.h>
+#include <app/CommandHandler.h>
+#include <app/EventLogging.h>
+#include <app/server/Server.h>
+#include <app/util/attribute-storage.h>
+#include <lib/support/CodeUtils.h>
+#include <lib/support/SortUtils.h>
+#include <lib/support/logging/CHIPLogging.h>
+#include <platform/CHIPDeviceLayer.h>
+
+#include <app-common/zap-generated/cluster-enums.h>
+
+#include <system/SystemClock.h>
+
+using namespace chip;
+using namespace chip::app;
+using namespace chip::DeviceLayer;
+using namespace chip::app::Clusters;
+using namespace chip::app::Clusters::TimeSynchronization;
+using namespace chip::app::Clusters::TimeSynchronization::Attributes;
+using chip::TimeSyncDataProvider;
+using chip::Protocols::InteractionModel::Status;
+
+// -----------------------------------------------------------------------------
+// Delegate Implementation
+
+namespace {
+
+Delegate * gDelegate = nullptr;
+
+Delegate * GetDelegate()
+{
+    if (gDelegate == nullptr)
+    {
+        static DefaultTimeSyncDelegate dg;
+        gDelegate = &dg;
+    }
+    return gDelegate;
+}
+} // namespace
+
+namespace chip {
+namespace app {
+namespace Clusters {
+namespace TimeSynchronization {
+
+void SetDefaultDelegate(Delegate * delegate)
+{
+    gDelegate = delegate;
+}
+
+Delegate * GetDefaultDelegate()
+{
+    return GetDelegate();
+}
+
+} // namespace TimeSynchronization
+} // namespace Clusters
+} // namespace app
+} // namespace chip
+
+constexpr uint64_t kChipEpochUsSinceUnixEpoch =
+    static_cast<uint64_t>(kChipEpochSecondsSinceUnixEpoch) * chip::kMicrosecondsPerSecond;
+
+static bool ChipEpochToUnixEpochMicro(uint64_t chipEpochTime, uint64_t & unixEpochTime)
+{
+    // in case chipEpochTime is too big and result overflows return false
+    if (chipEpochTime + kChipEpochUsSinceUnixEpoch < kChipEpochUsSinceUnixEpoch)
+    {
+        return false;
+    }
+    unixEpochTime = chipEpochTime + kChipEpochUsSinceUnixEpoch;
+    return true;
+}
+
+static bool UnixEpochToChipEpochMicro(uint64_t unixEpochTime, uint64_t & chipEpochTime)
+{
+    VerifyOrReturnValue(unixEpochTime >= kChipEpochUsSinceUnixEpoch, false);
+    chipEpochTime = unixEpochTime - kChipEpochUsSinceUnixEpoch;
+
+    return true;
+}
+
+static CHIP_ERROR UpdateUTCTime(uint64_t UTCTimeInChipEpochUs)
+{
+    System::Clock::Seconds32 lastKnownGoodChipEpoch;
+    uint64_t UTCTimeInUnixEpochUs;
+
+    ReturnErrorOnFailure(Server::GetInstance().GetFabricTable().GetLastKnownGoodChipEpochTime(lastKnownGoodChipEpoch));
+    VerifyOrReturnError(ChipEpochToUnixEpochMicro(UTCTimeInChipEpochUs, UTCTimeInUnixEpochUs), CHIP_ERROR_INVALID_TIME);
+
+    ReturnErrorOnFailure(Server::GetInstance().GetFabricTable().SetLastKnownGoodChipEpochTime(
```

Comment:
I don't think anything mandates that known good time cannot be set with older time than current. I had this in the code first but 1. nothing demands this and 2. it could actually be useful to correct a wrongly set good time.
Preventing Granularity going backward is done in `emberAfTimeSynchronizationClusterSetUTCTimeCallback`.

Reply:
3.5.6.1
```
A Node MAY adjust the Last Known Good UTC Time backwards if it believes the current Last Known Good UTC Time is incorrect and it has a good time value from a trusted source. The Node SHOULD NOT adjust the Last Known Good UTC to a time before the later of:

The build timestamp of its currently running software image

The not-before timestamp of any of its operational certificates (see [Section 6.4.5, “Node Operational Credentials Certificates”]).
```

So there are a couple of checks that could be added here or in the base command  handler.

Reply:
Tracked in https://github.com/project-chip/connectedhomeip/issues/27575, resolving this comment, followup should reference the issue.